### PR TITLE
cinny-desktop: init at 2.1.1

### PR DESCRIPTION
**Pull request by aveltras**
###### Description of changes

Added the desktop version of the [cinny](https://cinny.in/) matrix client.
This is an app built with [Tauri](https://github.com/tauri-apps/tauri).
I have used the debian package released on github (didn't manage to get the appimage version to run).
I have added a disableCompositing option to the package as the desktop window is blank is not doing this.
This seems to be an issue with Tauri apps when using nvidia.
Solves #172989

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -1257,6 +1257,12 @@
     githubId = 1217745;
     name = "Aldwin Vlasblom";
   };
+  aveltras = {
+    email = "romain.viallard@outlook.fr";
+    github = "aveltras";
+    githubId = 790607;
+    name = "Romain Viallard";
+  };
```
Comment (azahi) [inline maintainers/maintainer-list.nix:1265]: This should be a separate commit. Please consult [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) and don't forget about commit messages.

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  version = "2.0.4";
+
+  src = fetchurl {
+    url = "https://github.com/cinnyapp/cinny-desktop/releases/download/v${version}/cinny_${version}_amd64.deb";
+    sha256 = "sha256-udjLlYZTqOBTF9H7N6wJnIomFAeQvAczmEp3imE2IGY=";
+  };
+
+  nativeBuildInputs =
+    [ autoPatchelfHook dpkg gnused ]
+    ++ lib.optionals disableCompositing [ makeWrapper ];
```
Comment (azahi) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:25]:
```suggestion
  nativeBuildInputs = [
    autoPatchelfHook
    dpkg
    gnused
  ] ++ lib.optionals disableCompositing [
    makeWrapper
  ];
```

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+, disableCompositing ? false # https://bugs.webkit.org/show_bug.cgi?id=180739
+}:
+
+stdenv.mkDerivation rec {
+
+  name = "cinny-desktop";
+  version = "2.0.4";
+
+  src = fetchurl {
+    url = "https://github.com/cinnyapp/cinny-desktop/releases/download/v${version}/cinny_${version}_amd64.deb";
```
Comment (azahi) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:19]: Have you considered building it from scratch instead of repackaging a deb? It would be preferable to do so for Nixpkgs. If you still want to use debs, you must add this
```nix
sourceProvenance = with sourceTypes; [ binaryNativeCode ];
```
to `meta`.

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  nativeBuildInputs =
+    [ autoPatchelfHook dpkg gnused ]
+    ++ lib.optionals disableCompositing [ makeWrapper ];
+
+  buildInputs = [
+    openssl
+    webkitgtk
+  ];
+
+  unpackPhase = "true";
```
Comment (azahi) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:32]: This is enabled by default.
```suggestion
```

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 5]
```diff
@@ -0,0 +1,51 @@
+{ stdenv
+, lib
+, dpkg
+, fetchurl
+, autoPatchelfHook
+, openssl
+, gnused
+, webkitgtk
+, makeWrapper
+, disableCompositing ? false # https://bugs.webkit.org/show_bug.cgi?id=180739
```
Comment (azahi) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:10]: IMO this shouldn't be an option. A user could utilize `postFixup` to regenerate wrapper if he chooses to do so because having an NVIDIA card with proprietary drivers is a niche case. 

However, I would really like to hear a second opinion on this.

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  buildInputs = [
+    openssl
+    webkitgtk
+  ];
+
+  unpackPhase = "true";
+
+  installPhase = ''
+    mkdir -p $out/bin $out/share
+    dpkg -x $src $out/tmp
```
Comment (azahi) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:36]: This shouldn't be done at install phase. You can use `unpackCmd` instead.

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    sed -i 's/Exec=cinny/Exec=cinny-desktop/g' $out/tmp/usr/share/applications/cinny.desktop
```
Comment (azahi) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:37]: Please use `substituteInPlace`.

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    cp $out/tmp/usr/bin/cinny $out/bin/cinny-desktop
```
Comment (azahi) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:38]: Please use `intall -Dm755` to install executable. It would also be preferable to use it to install other file where possible.

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 9]
```diff
@@ -0,0 +1,51 @@
+{ stdenv
+, lib
+, dpkg
+, fetchurl
+, autoPatchelfHook
+, openssl
+, gnused
```
Comment (azahi) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:7]: `sed` is included in the `stdenv`.
```suggestion
```

**Review by azahi — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+  unpackPhase = "true";
+
+  installPhase = ''
+    mkdir -p $out/bin $out/share
+    dpkg -x $src $out/tmp
+    sed -i 's/Exec=cinny/Exec=cinny-desktop/g' $out/tmp/usr/share/applications/cinny.desktop
+    cp $out/tmp/usr/bin/cinny $out/bin/cinny-desktop
+    cp -r $out/tmp/usr/share/* $out/share/
+    ${if disableCompositing then "wrapProgram $out/bin/cinny-desktop --set WEBKIT_DISABLE_COMPOSITING_MODE 1" else ""}
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:40]:
```suggestion
    ${lib.optionalString disableCompositing "wrapProgram $out/bin/cinny-desktop --set WEBKIT_DISABLE_COMPOSITING_MODE 1"}
```

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix (hunk 6 above)
Comment (SuperSandro2000) [reply pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:36]: Also why is this first copied into tmp and later moved? We can copy this straight into place and change everything we want or delete unwanted files before copying

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix (hunk 5 above)
Comment (azahi) [reply pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:10]: @SuperSandro2000 could you please share your thoughts on this?

Comment (SuperSandro2000) [reply pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:10]: I don't think we are doing this anywhere else. I would recommend to set the environment variable in ``environment.variable`` to fix this globally.

**Comment by aveltras** [comment]
@azahi @SuperSandro2000 I have rebased with the edits.
Tauri doesn't seem to offer the possibility to build basic binaries but only AppImage or Debian package on Linux.

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix (hunk 3 above)
Comment (aveltras) [reply pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:16]: Tauri doesn't seem to offer the possibility to build basic binaries but only AppImage or Debian package on Linux.

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix (hunk 5 above)
Comment (aveltras) [reply pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix]: I have now removed it and set it in my global environment variables nixos setting.

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix (hunk 8 above)
Comment (aveltras) [reply pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix]: Not sure if this comment is still valid for the latest version of the PR.
I've taken inspiration from other builds in Nixpkgs now.

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+, dpkg
+, fetchurl
+, autoPatchelfHook
+, openssl
+, webkitgtk
+}:
+
+stdenv.mkDerivation rec {
+
+  name = "cinny-desktop";
```
Comment (azahi) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:12]:
```suggestion
stdenv.mkDerivation rec {
  name = "cinny-desktop";
```

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+  src = fetchurl {
+    url = "https://github.com/cinnyapp/cinny-desktop/releases/download/v${version}/cinny_${version}_amd64.deb";
+    sha256 = "sha256-udjLlYZTqOBTF9H7N6wJnIomFAeQvAczmEp3imE2IGY=";
+  };
+
+  nativeBuildInputs = [
+    autoPatchelfHook
+    dpkg
+  ];
+
```
Comment (azahi) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:24]:
```suggestion

  buildInputs = [
    openssl
    webkitgtk
  ];
```

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+  unpackCmd = "dpkg-deb -x $curSrc source";
+
+  installPhase = "mv usr $out";
+
+  buildInputs = [
+    openssl
+    webkitgtk
+  ];
```
Comment (azahi) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:32]:
```suggestion
```

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix (hunk 3 above)
Comment (azahi) [reply pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:16]: Okay then.

**Review by azahi — CHANGES_REQUESTED** [review]
Almost perfect. Just a couple of tiny nitpicks left.

**Comment by aveltras** [comment]
@azahi Should be good now.

**Review by azahi — CHANGES_REQUESTED** [review]

**Comment by aveltras** [comment]
@azahi now fixed, upstream has released a new version in the meantime so i changed it. Not sure if this changes anything regarding the current PR.

**Review by azahi — APPROVED** [review]
I'm going to approve this but for some reason I can't connect to any servers while, for instance, nheko works just fine. Could be something on my end, though.

image

It also prints out this while running:
```
[azahi@melian:~/.cache/nixpkgs-review/pr-185427-3]$ ./results/cinny-desktop/bin/cinny
Could not determine the accessibility bus address

(process:40015): GLib-GIO-CRITICAL **: 23:43:09.149: g_task_get_cancellable: assertion 'G_IS_TASK (task)' failed
```

**Comment by azahi** [comment]
Okay, `http://matrix.org` works but my homeserver `azahi.cc` doesn't. I guess this is some kind of an upstream issue or a misconfiguration on my part. 
image

**Comment by teutat3s** [comment]
> I'm going to approve this but for some reason I can't connect to any servers while, for instance, nheko works just fine. Could be something on my end, though.
> 
> image
> 
> It also prints out this while running:
> 
> ```
> [azahi@melian:~/.cache/nixpkgs-review/pr-185427-3]$ ./results/cinny-desktop/bin/cinny
> Could not determine the accessibility bus address
> 
> (process:40015): GLib-GIO-CRITICAL **: 23:43:09.149: g_task_get_cancellable: assertion 'G_IS_TASK (task)' failed
> ```

It seems there's an upstream issue for this https://github.com/cinnyapp/cinny/issues/693

**Review by teutat3s — APPROVED** [review]

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+  };
+
+  nativeBuildInputs = [
+    autoPatchelfHook
+    dpkg
+  ];
+
+  buildInputs = [
+    openssl
+    webkitgtk
```
Comment (teutat3s) [inline pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:26]:
```suggestion
    glib-networking
    openssl
    webkitgtk
    wrapGAppsHook
```

**Review by teutat3s — CHANGES_REQUESTED** [review]
Connection works fine for me with these changes

File: pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix (hunk 14 above)
Comment (aveltras) [reply pkgs/applications/networking/instant-messengers/cinny-desktop/default.nix:29]: I've pushed a rebase with your suggestions.

**Review by teutat3s — APPROVED** [review]
Perfect, works for me now

**Review by azahi — APPROVED** [review]
Works for me as well now.
